File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int[] parent;

    static int findSet(int x) {
        if (parent[x] != x) {
            return parent[x] = findSet(parent[x]);
        }
        return x;
    }
    static void union(int x, int y) {
        if (x > y) {
            parent[findSet(x)] = findSet(y);
        } else {
            parent[findSet(y)] = findSet(x);
        }

    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        parent = new int[n+1];
        for (int i = 1; i < n+1; i++) {
            parent[i] = i;
        }
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            union(a, b);
        }
        boolean[] visited = new boolean[n + 1];
        int ans = 0;
        for (int i = 1; i < n+1; i++) {
            if (visited[findSet(i)]) {
                continue;
            }
            visited[findSet(i)] = true;
            ans++;
        }
        System.out.println(ans);
        br.close();
    }
}

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
import sys

def find_set(a):
    if parent[a] != a:
        parent[a] = find_set(parent[a])
    return parent[a]       

def union(a, b):
    parent_a = parent[a]
    parent_b = parent[b]
    if parent_a == parent_b:
        return
    if parent_a < parent_b:
        parent[parent_b] = parent_a
    else:
        parent[parent_a] = parent_b      
        
n, m = map(int, sys.stdin.readline().split())
parent = [x for x in range(n + 1)]
for i in range(m):
    a, b = map(int, sys.stdin.readline().split())
    union(a, b)
visited = [False] * (n + 1)
ans = 0
for i in range(1, n+1):
    a = find_set(parent[i])
    if not visited[a]:
        ans += 1
        visited[a] = True

sys.stdout.write(str(ans))         